File: frontend/src/types/expense.ts
/**
 * Operating expenses (Martin M10).
 *
 * 🔴 Every `*_minor` field is INTEGER MINOR UNITS, like the rest of this
 * codebase: 850 is AED 8.50. Format with `formatMoney`, never by dividing here.
 *
 * `amount_minor` is the WHOLE invoice, VAT included. `tax_minor` is the part of
 * it that is VAT, and `net_minor` is the difference, computed on the server so
 * three screens cannot each arrive at their own answer.
 */

export type ExpenseStatus = "draft" | "unpaid" | "paid";

export interface ExpenseCategory {
  id: string;
  name: string;
  sort_order: number;
  is_active: boolean;
  notes: string | null;
  expense_count: number;
}

export interface ExpenseAttachment {
  id: string;
  media_id: string;
  filename: string | null;
  content_type: string;
  size_bytes: number;
  /**
   * Path on the API, not a public URL. It needs a bearer token, so it is
   * fetched as a blob rather than dropped into an `<img>` or an `<a href>`.
   */
  url: string;
}

export interface Expense {
  id: string;
  expense_date: string;
  payee: string;
  description: string | null;
  reference_number: string | null;
  amount_minor: number;
  tax_minor: number;
  net_minor: number;
  status: ExpenseStatus;
  payment_method: string | null;
  paid_on: string | null;
  notes: string | null;
  category_id: string | null;
  category_name: string | null;
  location_id: string | null;
  location_name: string | null;
  recorded_by: string | null;
  recorded_by_name: string | null;
  created_at: string;
  attachments: ExpenseAttachment[];
}

export interface ExpenseCreate {
  expense_date: string;
  payee: string;
  amount_minor: number;
  tax_minor?: number;
  category_id?: string | null;
  location_id?: string | null;
  description?: string | null;
  reference_number?: string | null;
  status?: ExpenseStatus;
  payment_method?: string | null;
  paid_on?: string | null;
  notes?: string | null;
}

export type ExpenseUpdate = Partial<ExpenseCreate>;

export interface ExpenseCategoryTotal {
  category_id: string | null;
  category_name: string;
  total_minor: number;
  expense_count: number;
}

export interface ExpenseSummary {
  date_from: string | null;
  date_to: string | null;
  location_id: string | null;
  total_minor: number;
  tax_total_minor: number;
  net_minor: number;
  unpaid_minor: number;
  expense_count: number;
  by_category: ExpenseCategoryTotal[];
}

File: backend/app/schemas/expense.py
"""Pydantic schemas for operating expenses (Martin M10).

🔴 Money is in MINOR UNITS, matching the inventory and procurement modules:
`amount_minor` of 850 is 8.50 AED. Nothing here multiplies or divides by 100.

🔴 No response model below carries a field default. A hand-built response with
defaults answers with a plausible wrong number when the builder forgets a field,
which is how the goods-receipt conversion reported 1 while the database held 400
(ERROR_LOG, 2026-09-04). A field that can genuinely be absent is typed `| None`
and passed explicitly.
"""

from __future__ import annotations

import uuid
from datetime import date, datetime
from decimal import Decimal
from typing import Annotated, Literal

from pydantic import BaseModel, Field, PlainSerializer

# Same reason as every other module in this codebase: Pydantic v2 serialises a
# `Decimal` to a JSON string, and the frontend types these as `number`.
Num = Annotated[
    Decimal,
    PlainSerializer(float, return_type=float, when_used="json"),
]

ExpenseStatus = Literal["draft", "unpaid", "paid"]


# ---------------------------------------------------------------------------
# CATEGORIES
# ---------------------------------------------------------------------------


class ExpenseCategoryCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=100)
    notes: str | None = None


class ExpenseCategoryUpdate(BaseModel):
    name: str | None = Field(None, min_length=1, max_length=100)
    sort_order: int | None = None
    is_active: bool | None = None
    notes: str | None = None


class ExpenseCategoryResponse(BaseModel):
    id: uuid.UUID
    name: str
    sort_order: int
    is_active: bool
    notes: str | None
    expense_count: int


# ---------------------------------------------------------------------------
# EXPENSES
# ---------------------------------------------------------------------------


class ExpenseCreate(BaseModel):
    expense_date: date
    payee: str = Field(..., min_length=1, max_length=200)
    amount_minor: Num = Field(..., ge=0)
    # The VAT contained IN `amount_minor`, not added to it. The service refuses
    # a tax larger than the total, and the database refuses it again.
    tax_minor: Num = Field(default=Decimal("0"), ge=0)
    category_id: uuid.UUID | None = None
    location_id: uuid.UUID | None = None
    description: str | None = None
    reference_number: str | None = Field(None, max_length=100)
    status: ExpenseStatus = "unpaid"
    payment_method: str | None = Field(None, max_length=50)
    paid_on: date | None = None
    notes: str | None = None


class ExpenseUpdate(BaseModel):
    expense_date: date | None = None
    payee: str | None = Field(None, min_length=1, max_length=200)
    amount_minor: Num | None = Field(None, ge=0)
    tax_minor: Num | None = Field(None, ge=0)
    category_id: uuid.UUID | None = None
    location_id: uuid.UUID | None = None
    description: str | None = None
    reference_number: str | None = Field(None, max_length=100)
    status: ExpenseStatus | None = None
    payment_method: str | None = Field(None, max_length=50)
    paid_on: date | None = None
    notes: str | None = None


class ExpenseAttachmentResponse(BaseModel):
    id: uuid.UUID
    media_id: uuid.UUID
    filename: str | None
    content_type: str
    size_bytes: int
    # Routed through the expenses router, not /media/{id}: an invoice for rent
    # or salaries needs a token, and /media is deliberately unauthenticated so
    # an <img> tag can fetch a menu photograph.
    url: str


class ExpenseResponse(BaseModel):
    id: uuid.UUID
    expense_date: date
    payee: str
    description: str | None
    reference_number: str | None
    amount_minor: Num
    tax_minor: Num
    net_minor: Num
    status: str
    payment_method: str | None
    paid_on: date | None
    notes: str | None
    category_id: uuid.UUID | None
    category_name: str | None
    location_id: uuid.UUID | None
    location_name: str | None
    recorded_by: uuid.UUID | None
    recorded_by_name: str | None
    created_at: datetime
    attachments: list[ExpenseAttachmentResponse]


# ---------------------------------------------------------------------------
# SUMMARY
# ---------------------------------------------------------------------------


class ExpenseCategoryTotal(BaseModel):
    category_id: uuid.UUID | None
    category_name: str
    total_minor: Num
    expense_count: int


class ExpenseSummaryResponse(BaseModel):
    date_from: date | None
    date_to: date | None
    location_id: uuid.UUID | None
    total_minor: Num
    tax_total_minor: Num
    net_minor: Num
    unpaid_minor: Num
    expense_count: int
    by_category: list[ExpenseCategoryTotal]
